pkg/controllers/workgenerator/controller.go

```diff
@@ -0,0 +1,329 @@
+	namespaceMatcher := client.InNamespace(fmt.Sprintf(utils.NamespaceNameFormat, resourceBinding.Spec.TargetCluster))
+	parentBindingLabelMatcher := client.MatchingLabels{
+		fleetv1beta1.ParentBindingLabel: resourceBinding.Name,
+	}
+	existingWork := make(map[string]*workv1alpha1.Work)
+	workList := &workv1alpha1.WorkList{}
+	if err := r.Client.List(ctx, workList, parentBindingLabelMatcher, namespaceMatcher); err != nil {
+		klog.ErrorS(err, "failed to list all the work associated with the resourceBinding", "resourceBinding", klog.KObj(resourceBinding))
+		return nil, controller.NewAPIServerError(true, err)
+	}
+	for _, work := range workList.Items {
+		existingWork[work.Name] = work.DeepCopy()
+	}
+	klog.InfoS("Get all the work associated", "numOfWork", len(existingWork), "resourceBinding", klog.KObj(resourceBinding))
+	return existingWork, nil
+}
+
+// syncAllWork generates all the work for the resourceBinding and apply them to the corresponding target cluster.
+func (r *Reconciler) syncAllWork(ctx context.Context, resourceBinding *fleetv1beta1.ClusterResourceBinding) (ctrl.Result, error) {
+	// Gather all the resource resourceSnapshots
+	resourceSnapshots, err := r.fetchAllResourceSnapshots(ctx, resourceBinding)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	// list all the corresponding works
+	works, err := r.listAllWorksAssociated(ctx, resourceBinding)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// create/update the corresponding work for each snapshot
+	activeWork := make(map[string]bool, len(resourceSnapshots))
+	for _, snapshot := range resourceSnapshots {
+		workName := getWorkNameFromSnapshotName(snapshot.Name)
+		activeWork[workName] = true
+		// TODO(RZ): issue those requests in parallel to speed up the process
+		err = r.upsertWork(ctx, works[workName], snapshot, resourceBinding)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+	}
+
+	//  delete the works that are not associated with any resource snapshot
+	for _, work := range works {
+		if activeWork[work.Name] {
+			continue
+		}
+		klog.V(2).InfoS("Delete the work that is not associated with any resource snapshot", "work", klog.KObj(work))
+		if err := r.Client.Delete(ctx, work); err != nil {
+			klog.ErrorS(err, "failed to delete the no longer needed work", "work", klog.KObj(work))
+			if !errors.IsNotFound(err) {
+				return ctrl.Result{}, controller.NewAPIServerError(false, err)
+			}
+		}
+	}
+
+	klog.V(2).InfoS("Successfully synced all the work associated with the resourceBinding", "resourceBinding", klog.KObj(resourceBinding))
+	return ctrl.Result{}, nil
+}
+
+// fetchAllResourceSnapshots gathers all the resource snapshots for the resource binding.
+func (r *Reconciler) fetchAllResourceSnapshots(ctx context.Context, resourceBinding *fleetv1beta1.ClusterResourceBinding) (map[string]*fleetv1beta1.ClusterResourceSnapshot, error) {
+	resourceSnapshots := make(map[string]*fleetv1beta1.ClusterResourceSnapshot, 1)
```

Comment:
just curious why we need the map here? array is good enough?

Reply:
map is the easiest way to dedup since the "master" snapshot will be included in the following list operation again.

Reply:
oh i see, forget we'll put the first snapshot twice.